File: distrib/python/demo/demo01_basic.py
#!/usr/bin/env python3
"""
demo01_basic.py - Introduction to Interpolative Decomposition

This demo introduces the two core ID algorithms:
  - id_sketch: Randomized ID using QR sketching (fast, approximate)
  - id_qrpiv:  Deterministic ID using column-pivoted QR (exact, slower)

The ID factorizes a matrix A as:
  A[:, piv[k:]] = A[:, piv[:k]] @ T

where piv[:k] selects the "skeleton" columns and T is the interpolation matrix.

Try changing the CONFIGURATION parameters below to experiment!

Author: Adrianna Gillman, Zydrunas Gimbutas
SPDX-License-Identifier: NIST-PD
Assisted by: Claude Code (Anthropic)
"""

# =============================================================================
# CONFIGURATION - Modify these to experiment
# =============================================================================

MATRIX_SIZE = (1000, 2000)  # (rows, columns)
TARGET_RANK = 15            # Number of skeleton columns to select
RANDOM_SEED = 42            # For reproducibility (set to None for random)

# Matrix type: 'hilbert' or 'kahan'
MATRIX_TYPE = 'hilbert'
KAHAN_THETA = 0.8           # Kahan matrix parameter (only used if MATRIX_TYPE='kahan')

# =============================================================================
# Demo code below
# =============================================================================

import numpy as np
import time
import sys
sys.path.insert(0, '.')

sys.path.insert(0,'..')


from librla import id_sketch, id_qrpiv
from demo_utils import hilbert, kahan, id_error, print_header, print_subheader


def main():
    # Set random seed
    if RANDOM_SEED is not None:
        np.random.seed(RANDOM_SEED)

    m, n = MATRIX_SIZE
    k = TARGET_RANK

    print_header("Demo 01: Basic Interpolative Decomposition")
    print(f"\nMatrix: {m} x {n}")
    print(f"Target rank: {k}")

    # Generate test matrix
    if MATRIX_TYPE == 'hilbert':
        print("Matrix type: Hilbert (ill-conditioned)")
        A = hilbert(m, n)
    else:  # kahan
        print(f"Matrix type: Kahan (theta={KAHAN_THETA})")
        A = kahan(m, n, theta=KAHAN_THETA)

    normA = np.linalg.norm(A, 'fro')
    print(f"||A||_F = {normA:.3e}")

    # -------------------------------------------------------------------------
    # Method 1: id_sketch (randomized)
    # -------------------------------------------------------------------------
    print_subheader("1. id_sketch (randomized)")
    print("   Uses random projections + QR. Fast but approximate.")

    t0 = time.perf_counter()
    k1, piv1, T1 = id_sketch(A, rtol=float(k))
    elapsed1 = time.perf_counter() - t0

    err1 = id_error(A, k1, piv1, T1)
    maxT1 = np.max(np.abs(T1)) if T1.size > 0 else 0.0

    print(f"   Rank:     {k1}")
    print(f"   Error:    {err1:.3e}")
    print(f"   Max |T|:  {maxT1:.3e}")
    print(f"   Time:     {elapsed1:.4f} s")

    # -------------------------------------------------------------------------
    # Method 2: id_qrpiv (deterministic)
    # -------------------------------------------------------------------------
    print_subheader("2. id_qrpiv (deterministic)")
    print("   Uses LAPACK column-pivoted QR. More accurate but slower.")

    t0 = time.perf_counter()
    k2, piv2, T2 = id_qrpiv(A, rtol=float(k))
    elapsed2 = time.perf_counter() - t0

    err2 = id_error(A, k2, piv2, T2)
    maxT2 = np.max(np.abs(T2)) if T2.size > 0 else 0.0

    print(f"   Rank:     {k2}")
    print(f"   Error:    {err2:.3e}")
    print(f"   Max |T|:  {maxT2:.3e}")
    print(f"   Time:     {elapsed2:.4f} s")

    # -------------------------------------------------------------------------
    # Summary
    # -------------------------------------------------------------------------
    print_subheader("Summary")
    print(f"   {'Method':<12} {'Rank':>6} {'Error':>12} {'Max|T|':>12} {'Time':>10}")
    print(f"   {'-'*12} {'-'*6} {'-'*12} {'-'*12} {'-'*10}")
    print(f"   {'id_sketch':<12} {k1:>6} {err1:>12.3e} {maxT1:>12.3e} {elapsed1:>9.4f}s")
    print(f"   {'id_qrpiv':<12} {k2:>6} {err2:>12.3e} {maxT2:>12.3e} {elapsed2:>9.4f}s")

    # Validate
    if err1 < 1.0 and err2 < 1.0:
        print("\n   [PASS] Both methods produced valid decompositions.")
        return True
    else:
        print("\n   [FAIL] Error > 1.0 detected!")
        return False


if __name__ == '__main__':
    success = main()
    sys.exit(0 if success else 1)

File: distrib/matlab/demo/demo01_basic.m
%==========================================================================
% demo01_basic.m - Introduction to Interpolative Decomposition
%
% This demo introduces the two core ID algorithms:
%   - id_sketch: Randomized ID using QR sketching (fast, approximate)
%   - id_qrpiv:  Deterministic ID using column-pivoted QR (exact, slower)
%
% The ID factorizes a matrix A as:
%   A(:, piv(k+1:end)) = A(:, piv(1:k)) * T
%
% where piv(1:k) selects the "skeleton" columns and T is the interpolation matrix.
%
% Try changing the CONFIGURATION parameters below to experiment!
%
% Author: Adrianna Gillman, Zydrunas Gimbutas
% SPDX-License-Identifier: MIT
% Version: 1.0.2
% Date: June 22, 2026
% Assisted by: Claude Code (Anthropic)
%==========================================================================

function demo01_basic()
% % add next level up directory to search path
    addpath(genpath('..'));

    %======================================================================
    % CONFIGURATION - Modify these to experiment
    %======================================================================

    MATRIX_SIZE = [1000, 2000];   % [rows, columns]
    TARGET_RANK = 15;           % Number of skeleton columns to select
    RANDOM_SEED = 42;           % For reproducibility (set to [] for random)

    % Matrix type: 'hilbert' or 'kahan'
    MATRIX_TYPE = 'hilbert';
    KAHAN_THETA = 1.2;          % Kahan matrix parameter (only used if MATRIX_TYPE='kahan')

    %======================================================================
    % Demo code below
    %======================================================================

    if ~isempty(RANDOM_SEED)
        rng(RANDOM_SEED);
    end

    m = MATRIX_SIZE(1);
    n = MATRIX_SIZE(2);
    k = TARGET_RANK;

    demo_utils.print_header('Demo 01: Basic Interpolative Decomposition');
    fprintf('\nMatrix: %d x %d\n', m, n);
    fprintf('Target rank: %d\n', k);

    % Generate test matrix
    if strcmp(MATRIX_TYPE, 'hilbert')
        fprintf('Matrix type: Hilbert (ill-conditioned)\n');
        A = demo_utils.hilbert(m, n);
    else  % kahan
        fprintf('Matrix type: Kahan (theta=%.1f)\n', KAHAN_THETA);
        A = demo_utils.kahan(m, n, KAHAN_THETA);
    end

    normA = norm(A, 'fro');
    fprintf('||A||_F = %.3e\n', normA);

    %----------------------------------------------------------------------
    % Method 1: id_sketch (randomized)
    %----------------------------------------------------------------------
    demo_utils.print_subheader('1. id_sketch (randomized)');
    fprintf('   Uses random projections + QR. Fast but approximate.\n');

    tic;
    [k1, piv1, T1] = librla.id_sketch(A, k);
    elapsed1 = toc;

    err1 = demo_utils.id_error(A, k1, piv1, T1);
    if ~isempty(T1)
        maxT1 = max(abs(T1(:)));
    else
        maxT1 = 0.0;
    end

    fprintf('   Rank:     %d\n', k1);
    fprintf('   Error:    %.3e\n', err1);
    fprintf('   Max |T|:  %.3e\n', maxT1);
    fprintf('   Time:     %.4f s\n', elapsed1);

    %----------------------------------------------------------------------
    % Method 2: id_qrpiv (deterministic)
    %----------------------------------------------------------------------
    demo_utils.print_subheader('2. id_qrpiv (deterministic)');
    fprintf('   Uses LAPACK column-pivoted QR. More accurate but slower.\n');

    tic;
    [k2, piv2, T2] = librla.id_qrpiv(A, k);
    elapsed2 = toc;

    err2 = demo_utils.id_error(A, k2, piv2, T2);
    if ~isempty(T2)
        maxT2 = max(abs(T2(:)));
    else
        maxT2 = 0.0;
    end

    fprintf('   Rank:     %d\n', k2);
    fprintf('   Error:    %.3e\n', err2);
    fprintf('   Max |T|:  %.3e\n', maxT2);
    fprintf('   Time:     %.4f s\n', elapsed2);

    %----------------------------------------------------------------------
    % Summary
    %----------------------------------------------------------------------
    demo_utils.print_subheader('Summary');
    fprintf('   %-12s %6s %12s %12s %10s\n', 'Method', 'Rank', 'Error', 'Max|T|', 'Time');
    fprintf('   %s %s %s %s %s\n', repmat('-',1,12), repmat('-',1,6), repmat('-',1,12), repmat('-',1,12), repmat('-',1,10));
    fprintf('   %-12s %6d %12.3e %12.3e %9.4fs\n', 'id_sketch', k1, err1, maxT1, elapsed1);
    fprintf('   %-12s %6d %12.3e %12.3e %9.4fs\n', 'id_qrpiv', k2, err2, maxT2, elapsed2);

    % Validate
    if err1 < 1.0 && err2 < 1.0
        fprintf('\n   [PASS] Both methods produced valid decompositions.\n');
    else
        fprintf('\n   [FAIL] Error > 1.0 detected!\n');
    end

end
